### PR TITLE
Hat snapshots

### PR DESCRIPTION
**Pull request by pokey**
### Description

Allows voice engine to emit a signal to indicate start of phrase, which causes the hat token map to be frozen during the course of all commands issued during the given phrase.  The ranges will be updated as the document changes, but the same hat names will refer to the same logical document token during the course of the phrase.  This feature allows users to issue longer voice commands without worrying about the hats shifting during the course of phrase execution

### Related issues

- https://github.com/pokey/cursorless-talon/pull/101
- https://github.com/knausj85/knausj_talon/pull/676
- https://github.com/pokey/command-server/pull/10

### Todo
- [x] Add transformation that can upgrade old tests
- [x] Improve documentation for transform script
- [x] Record video walkthrough of using transform script
- [x] Add code links to hat snapshots doc (will do after merge)
- [x] Figure out how to mock prePhrase signal
- [x] Don't try to touch signal from talon if communication dir doesn't exist
- [x] Handle case where user has this version of cursorless-talon but old command-client; need to fall back to not using snapshot?  Do this talon side or extension side?
- [x] Check age of snapshot if it is requested. If too old, show error message and don't use snapshot
- [x] Switch `signals` to `inboundSignals`
- [x] Remove `getNamedSignal` from command-server API (here and in extension)
- [x] Fix issue with toggle decorations
- [x] Figure out why seeing duplicate hats after clone line then undo
- [x] Figure out promise rejection undefined

### CODE REVIEW

File: src/core/commandRunner/types.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { ActionType, PartialTarget } from "../../typings/Types";
```
Comment (pokey) [inline src/core/commandRunner/types.ts:1]: Here's the new command argument type definitions

File: src/core/commandRunner/CommandRunner.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      vscode.window.showErrorMessage(err.message);
+      console.debug(err.message);
+      console.debug(err.stack);
+      throw err;
+    }
+  }
+
+  private runCommandBackwardCompatible(
+    spokenFormOrCommandArgument: string | CommandArgument,
+    ...rest: unknown[]
```
Comment (pokey) [inline src/core/commandRunner/CommandRunner.ts:121]: Here's the ugliness that allows us to handle either old or new command arguments

File: src/test/suite/fixtures/recorded/actions/bringAirAndBatAndCap.yml [hunk 3]
```diff
@@ -1,8 +1,9 @@
-spokenForm: bring air and bat and cap
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/actions/bringAirAndBatAndCap.yml:1]: New recorded test syntax just stores an exact snapshot of the command argument that comes in to the cursorless command

File: src/util/canonicalizeAndValidateCommand.ts [hunk 4]
```diff
@@ -1,23 +1,41 @@
...
+ *
+ * @param commandArgument The command argument to normalize
+ * @returns The normalized command argument
+ */
 export function canonicalizeAndValidateCommand(
-  inputActionName: string,
-  inputPartialTargets: PartialTarget[],
-  inputExtraArgs: any[]
-) {
+  commandArgument: CommandArgument
```
Comment (pokey) [inline src/util/canonicalizeAndValidateCommand.ts:18]: Now that we have one argument packaged together, we can just input a command argument and return a new command argument that is latest version

File: src/testUtil/TestCase.ts [hunk 5]
```diff
@@ -10,25 +10,21 @@ import { marksToPlainObject, SerializedMarks } from "./toPlainObject";
...
 import { pick } from "lodash";
+import { ReadOnlyHatMap } from "../core/IndividualHatMap";
+import { CommandArgument } from "../core/commandRunner/types";
 
-export type TestCaseCommand = {
-  actionName: ActionType;
-  partialTargets: PartialTarget[];
-  extraArgs: any[];
-};
+export type TestCaseCommand = CommandArgument;
```
Comment (pokey) [inline src/testUtil/TestCase.ts:16]: Notice how this one is just a regular command argument now

File: src/core/Decorations.ts [hunk 6]
```diff
@@ -146,9 +191,9 @@ export default class Decorations {
     shapePenalties.default = 0;
     colorPenalties.default = 0;
 
-    const activeHatColors = HAT_COLORS.filter(
-      (color) => colorEnablement[color]
-    );
+    const activeHatColors = isTesting()
+      ? HAT_COLORS
+      : HAT_COLORS.filter((color) => colorEnablement[color]);
```
Comment (pokey) [inline src/core/Decorations.ts:196]: So that unit tests don't fail locally if you have some colors disabled

Comment (AndreasArvidsson) [reply src/core/Decorations.ts:196]: That might warrant a comment in the code?

File: src/core/commandRunner/CommandRunner.ts (hunk 2 above)
Comment (AndreasArvidsson) [reply src/core/commandRunner/CommandRunner.ts:121]: I think that works fine

File: src/core/Decorations.ts [hunk 7]
```diff
@@ -30,24 +30,60 @@ export interface NamedDecoration {
...
+  private decorationChangeListeners: DecorationChangeListener[] = [];
+  private disposables: vscode.Disposable[] = [];
 
   constructor(private graph: Graph) {
-    this.constructDecorations(graph.fontMeasurements);
     graph.extensionContext.subscriptions.push(this);
+
+    this.recomputeDecorationStyles = this.recomputeDecorationStyles.bind(this);
+
+    this.disposables.push(
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply src/core/Decorations.ts:49]: Looks good really clean things up

File: src/test/suite/fixtures/recorded/actions/bringAirAndBatAndCap.yml (hunk 3 above)
Comment (AndreasArvidsson) [reply src/test/suite/fixtures/recorded/actions/bringAirAndBatAndCap.yml:1]: Sounds good. Really inline with our discussion about that the tests should only capture what the user actually said.

**Comment by AndreasArvidsson** [comment]
I must admit I haven't looked through all the code. But I think the changes to the api and command version looks good.

**Review by AndreasArvidsson — APPROVED** [review]

File: src/core/Decorations.ts (hunk 6 above)
Comment (pokey) [reply src/core/Decorations.ts:196]: Marking this one unresolved until I have a chance to add the comment

**Pull request merged**